Style engine: migrate functions, classes and tests

ramonjd opened the pull request:
This PR migrates the Style Engine PHP functions, classes and tests into Core for 6.1.

It reflects the latest changes from https://github.com/WordPress/gutenberg/pull/43840

See tracking issue: https://github.com/WordPress/gutenberg/issues/43440

## Backport PRs blocked by this changeset (that I know of)

- https://github.com/WordPress/wordpress-develop/pull/3218
- https://github.com/WordPress/wordpress-develop/pull/3203
- https://github.com/WordPress/wordpress-develop/pull/3204

Trac ticket: https://core.trac.wordpress.org/ticket/56467

ramonjd committed:
Initial commit:
- Migration of the style engine classes and unit tests from Gutenberg to Core.

ramonjd committed:
Adding @covers annotations

ramonjd committed:
Changes from https://github.com/WordPress/gutenberg/pull/43886

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+/**
+ * Tests for registering, storing and generating CSS declarations.
+ */
+class WP_Style_Engine_CSS_Declarations_Test extends WP_UnitTestCase {
```
```suggestion
class Tests_Style_Engine_wpStyleEngineCSSDeclarations extends WP_UnitTestCase {
```
See [Core Handbook - Writing PHPUnit Tests - Test Classes](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#test-classes). 

Note: Despite this, `wp` is usually lowercase in the test suite. For consistency, the suggestion maintains the lowercase `wp` so that potential future changes to this in the test suite can be made in one go.

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+ */
+
+/**
+ * Tests for registering, storing and generating CSS declarations.
```
```suggestion
 * Tests for registering, storing and generating CSS declarations.
 *
 * @coversDefaultClass WP_Style_Engine_CSS_Declarations 
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+ */
+class WP_Style_Engine_CSS_Declarations_Test extends WP_UnitTestCase {
+	/**
+	 * Should set declarations on instantiation.
```
```suggestion
	 * Tests that declarations are set on instantiation.
	 *
	 * @ticket 55966
	 *
	 * @covers ::__construct
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should set declarations on instantiation.
+	 */
+	public function test_instantiate_with_declarations() {
```
```suggestion
	public function test_should_set_declarations_on_instantiation() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Should add declarations.
```
```suggestion
	 * Tests that declarations are added.
	 * 
	 * @ticket 55966
	 * 
	 * @covers ::add_declarations
	 * @covers ::add_declaration
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should add declarations.
+	 */
+	public function test_add_declarations() {
```
```suggestion
	public function test_should_add_declarations() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Should add declarations.
```
```suggestion
	 * Tests that new declarations are added to existing declarations.
	 * 
	 * @ticket 55966
	 * 
	 * @covers ::add_declarations
	 * @covers ::add_declaration
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should add declarations.
+	 */
+	public function test_add_a_single_declaration() {
```
```suggestion
	public function test_should_add_new_declarations_to_existing_declarations() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Should sanitize properties before storing.
```
```suggestion
	 * Tests that properties are sanitized before being stored.
	 * 
	 * @ticket 55966
	 * 
	 * @covers ::sanitize_property
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should sanitize properties before storing.
+	 */
+	public function test_sanitize_properties() {
```
```suggestion
	public function test_should_sanitize_properties_before_storing() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+		$this->assertSame(
+			'color:red;border-top-left-radius:99px;text-decoration:underline;',
+			$css_declarations->get_declarations_string()
+		);
```
```suggestion
	/**
	 * Tests that CSS declarations are compiled into a CSS declarations block string.
	 *
	 * @ticket 55966
	 *
	 * @covers ::get_declarations_string
	 * 
	 * @dataProvider data_should_compile_css_declarations_to_css_declarations_string
	 *
	 * @param string $expected        The expected declarations block string.
	 * @param bool   $should_prettify Optional. Whether to pretty the string. Default false.
	 * @param int    $indent_count    Optional. The number of tab indents. Default false.
	 */
	public function test_should_compile_css_declarations_to_css_declarations_string( $expected, $should_prettify = false, $indent_count = 0 ) {
		$input_declarations = array(
			'color'                  => 'red',
			'border-top-left-radius' => '99px',
			'text-decoration'        => 'underline',
		);
		$css_declarations   = new WP_Style_Engine_CSS_Declarations( $input_declarations );

		$this->assertSame(
			$expected,
			$css_declarations->get_declarations_string( $should_prettify, $indent_count )
		);
```

This test should use a data provider accepting  `$expected`, `$should_prettify = false`, `$indent_count = 0`.

From there, the following data provider can be used to replace the next three test methods, and add more datasets without being _required_ to maintain new test methods.

```php
	/**
	 * Data provider.
	 *
	 * @return array
	 */
	public function data_should_compile_css_declarations_to_css_declarations_string() {
		return array(
			'unprettified, no indent'  => array(
				'expected' => 'color:red;border-top-left-radius:99px;text-decoration:underline;',
			),
			'unprettified, one indent' => array(
				'expected'        => 'color:red;border-top-left-radius:99px;text-decoration:underline;',
				'should_prettify' => false,
				'indent_count'    => 1,
			),
			'prettified, no indent'    => array(
				'expected'        => 'color: red; border-top-left-radius: 99px; text-decoration: underline;',
				'should_prettify' => true,
			),
			'prettified, one indent'   => array(
				'expected'        => "\tcolor: red;\n\tborder-top-left-radius: 99px;\n\ttext-decoration: underline;",
				'should_prettify' => true,
				'indent_count'    => 1,
			),
			'prettified, two indents'  => array(
				'expected'        => "\t\tcolor: red;\n\t\tborder-top-left-radius: 99px;\n\t\ttext-decoration: underline;",
				'should_prettify' => true,
				'indent_count'    => 2,
			),
		);
	}
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Should escape values and run the CSS through safecss_filter_attr.
```
```suggestion
	 * Tests that values are escaped and run the CSS through safecss_filter_attr().
	 *
	 * @ticket 55966
	 *
	 * @covers ::filter_declaration
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+		$this->assertSame(
+			'margin-right:10em;',
+			$css_declarations->get_declarations_string()
+		);
```
```suggestion
		$mock_action = new MockAction();
		add_filter( 'safe_style_css', array( $mock_action, 'filter' );
		
		$this->assertSame(
			1,
			$mock_action->get_call_count(),
			'"safe_style_css" filters were not applied.'
		);
		
		$this->assertSame(
			'margin-right:10em;',
			$css_declarations->get_declarations_string(),
			'Unsafe properties and values were not removed'
		);
```
Part of this tests that the returned value has been filtered correctly. While the current implementation uses `safecss_filter_attr()`, a future change could find another approach to achieve the same result.

As `safecss_filter_attr()` applies `safe_style_css` filters, such a change would break BC if `safe_style_css` filters may not be applied in a new implementation.

To protect BC, this should also test that `safe_style_css` filters have been run, as shown above.

Additionally, the current test should be evaluated to check if `safecss_filter_attr_allow_css` filters currently run. If so, this should also be covered.

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should escape values and run the CSS through safecss_filter_attr.
+	 */
+	public function test_remove_unsafe_properties_and_values() {
```
```suggestion
	public function test_should_remove_unsafe_properties_and_values() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Should allow calc, clamp, min, max, and minmax CSS functions.
```
```suggestion
	 * Tests that only calc, clamp, min, max, and minmax CSS functions are allowed.
	 *
	 * @ticket 55966
	 *
	 * @covers ::filter_declaration
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should allow calc, clamp, min, max, and minmax CSS functions.
+	 */
+	public function test_allow_particular_css_functions() {
```
```suggestion
	public function test_should_allow_particular_css_functions() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Should remove a declaration
```
```suggestion
	 * Tests that a single declaration is removed.
	 *
	 * @ticket 55966
	 *
	 * @covers ::remove_declaration
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should remove a declaration
+	 */
+	public function test_remove_declaration() {
```
```suggestion
	public function test_should_remove_a_single_declaration() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+		$this->assertSame(
+			'color:tomato;margin:10em 10em 20em 1px;font-family:Happy Font serif;',
+			$css_declarations->get_declarations_string()
+		);
```
```suggestion
		$this->assertSame(
			'color:tomato;margin:10em 10em 20em 1px;font-family:Happy Font serif;',
			$css_declarations->get_declarations_string(),
			'Declarations were not set on instantiation'
		);
```
Test methods with multiple assertions need a `$message` parameter for each assertion to explain what went wrong. See [Core Handbook - Writing PHPUnit Tests - Using Assertions](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions).

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+		$this->assertSame(
+			'margin:10em 10em 20em 1px;font-family:Happy Font serif;',
+			$css_declarations->get_declarations_string()
+		);
```
```suggestion
		$this->assertSame(
			'margin:10em 10em 20em 1px;font-family:Happy Font serif;',
			$css_declarations->get_declarations_string(),
			'The "color" declaration was not removed'
		);
```
See above.

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Should remove declarations
```
```suggestion
	 * Tests that multiple declarations are removed.
	 *
	 * @ticket 55966
	 *
	 * @covers ::remove_declarations
	 * @covers ::remove_declaration
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+	/**
+	 * Should remove declarations
+	 */
+	public function test_remove_declarations() {
```
```suggestion
	public function test_should_remove_multiple_declarations() {
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+		$this->assertSame(
+			'color:cucumber;margin:10em 10em 20em 1px;font-family:Happy Font serif;',
+			$css_declarations->get_declarations_string()
+		);
```
```suggestion
		$this->assertSame(
			'color:cucumber;margin:10em 10em 20em 1px;font-family:Happy Font serif;',
			$css_declarations->get_declarations_string(),
			'Declarations were not set on instantiation'
		);
```
See above.

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssDeclarations.php:
```diff
@@ -0,0 +1,235 @@
+		$this->assertSame(
+			'font-family:Happy Font serif;',
+			$css_declarations->get_declarations_string()
+		);
```
```suggestion
		$this->assertSame(
			'font-family:Happy Font serif;',
			$css_declarations->get_declarations_string(),
			'The "color" and "margin" declarations were not removed'
		);
```
See above.

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -0,0 +1,154 @@
+	 * @param array $options array(
+	 *    'optimize' => (boolean) Whether to optimize the CSS output, e.g., combine rules.
+	 *    'prettify' => (boolean) Whether to add new lines to output.
+	 * );.
```
```suggestion
	 * @param array $options {
	 *     Optional. Formatting options. Default empty array.
	 *
	 *     @type bool $optimize Whether to optimize the CSS output, e.g., combine rules.
	 *     @type bool $prettify Whether to add new lines to output.
	 * }
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -0,0 +1,154 @@
+	 *
+	 * @return WP_Style_Engine_Processor Returns the object to allow chaining methods.
+	 */
+	public function add_store( WP_Style_Engine_CSS_Rules_Store $store ) {
```
Core typically doesn't provide a parameter type.

If this _is_ removed, type checking should be done via `if ( ! $store instanceof WP_Style_Engine_CSS_Rules ) {}`, a `_doing_it_wrong()` thrown, and the `@return` type updated to include an appropriate return type on failure.

However, pinging @SergeyBiryukov for a second opinion on this.

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	/**
+	 * Get the CSS rules as a string.
```
```suggestion
	 * Gets the CSS rules as a string.
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -0,0 +1,154 @@
+	protected $css_rules = array();
+
+	/**
+	 * Add a store to the processor.
```
```suggestion
	 * Adds a store to the processor.
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -0,0 +1,154 @@
+class WP_Style_Engine_Processor {
+
+	/**
+	 * The Style-Engine Store objects
```
```suggestion
	 * The Style-Engine Store objects.
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -0,0 +1,154 @@
+		if ( ! is_array( $css_rules ) ) {
+			$css_rules = array( $css_rules );
+		}
+		foreach ( $css_rules as $rule ) {
```
```suggestion
		}

		foreach ( $css_rules as $rule ) {
```

costdev commented on src/wp-includes/style-engine.php:
```diff
@@ -0,0 +1,150 @@
+ *                  When set, the style engine will attempt to store the CSS rules, where a selector is also passed.
+ *     'convert_vars_to_classnames' => (boolean) Whether to skip converting CSS var:? values to var( --wp--preset--* ) values. Default is `false`.
+ *     'selector'                   => (string) When a selector is passed, `generate()` will return a full CSS rule `$selector { ...rules }`, otherwise a concatenated string of properties and values.
+ * );.
```
```suggestion
 * @param array $options {
 *     Optional. An array of options. Default empty array.
 *
 *     @type string|null $context                    An identifier describing the origin of the style object, such as 'block-supports' or 'global-styles'. 
 *                                                   When set, the style engine will attempt to store the CSS rules, where a selector is also passed.
 *                                                   Default 'block-supports'.
 *     @type bool        $convert_vars_to_classnames Whether to skip converting CSS var:? values to var( --wp--preset--* ) values. Default `false`.
 *     @type string      $selector                   When a selector is passed, generate() will return a full CSS rule `$selector { ...rules }`,
 *                                                   otherwise a concatenated string of properties and values.
 * }
```
This should also be done for all expanded `array` parameter documentation.


Additional Query: In "Whether to skip converting CSS var:? values", what's the `:?`?

ramonjd replied:
👍 

Good catch. It should be `var:preset|<PRESET_TYPE>|<PRESET_SLUG>`. I'll update all instances.

costdev commented on tests/phpunit/tests/style-engine/styleEngine.php:
```diff
@@ -0,0 +1,668 @@
+	 * Tear down after each test.
+	 */
+	public function tear_down() {
+		parent::tear_down();
```
`parent::tear_down()` should be the last line of the `tear_down()` method.

costdev commented on tests/phpunit/tests/style-engine/styleEngine.php:
```diff
@@ -0,0 +1,668 @@
+	 *
+	 * @dataProvider data_get_styles_fixtures
+	 *
+	 * @param array  $block_styles The incoming block styles object.
```
Parameter names should be in line with each other, and their descriptions should be in line with each other. See some other tests in the Core test suite for examples.

costdev reviewed (changes requested):
As this review is becoming quite large and the comments are quite similar, I'll submit it now.

The feedback on the tests applies to all other test files, as do documentation comments in source files.

A couple of general things:
- Test class names should have the format `Tests_Folder_Name_ThingInTitleCase`.
- Ensure all test methods have the `@ticket` annotation.
- Ensure all test methods have the `@covers` (or `@coversNothing` if it doesn't cover something) annotations, as this will allow us all to see where coverage may be great/need improvement. 
- As shown in the review, `@coversDefaultClass` can be used in the class docblock for a file that tests a single class, which allows for shorter `covers` annotations using simply the method name like `::method_name`.
- Where possible, try to start test methods with `test_should_` / `test_should_not`. Using `test_method_name` or `test_function_name` is typically only useful for small source functions, or for test methods that use a data provider.

ramonjd committed:
Formatting PHP doc comments and params that describe associative arrays

Updating test function names and annotations.

ramonjd committed:
formatting

ramonjd committed:
Combining tests for get_declarations_string
Updating php doc params

ramonjd committed:
Adding message parameter to assertions where more than one assertion is used in a test method.

ramonjd committed:
Updating PHP docs
Testing for WP_Style_Engine_CSS_Rules_Store in WP_Style_Engine_Processor->add_store

ramonjd committed:
Refactoring sanitization tests for CSS declarations
Counting calls to mock filters executed in safecss_filter_attr() (kses.php)

ramonjd commented:
Thanks @costdev for volunteering and taking the time to review the code, and also for explaining the required changes 🙇 

I think I've covered just about everything, but will take another sweep over later to spot if I've missed something.

The final failing test should pass once https://core.trac.wordpress.org/ticket/55966 has been committed.

ramonjd committed:
Formatting styleEngine tests
Removing duplicated test (we were already testing context below)

ramonjd committed:
My blod is typo

ramonjd committed:
Imperative < active voice

ramonjd committed:
Remove Gutenberg suffix

ramonjd committed:
renaming test group suite to kebab case

costdev commented on src/wp-includes/style-engine.php:
```diff
@@ -0,0 +1,164 @@
+ * }
+ */
+function wp_style_engine_get_styles( $block_styles, $options = array() ) {
+	if ( ! class_exists( 'WP_Style_Engine' ) ) {
```
Quick check: In what scenario would this class not exist in Core, or is this something for the plugin but not for Core?

ramonjd replied:
This was a carry over from Gutenberg where compat makes one defensive. Agreed it's not required here.

costdev commented on src/wp-includes/style-engine.php:
```diff
@@ -0,0 +1,164 @@
+ * $styles = wp_style_engine_get_styles( array( 'color' => array( 'text' => '#cccccc' ) ) );
+ * // Returns `array( 'css' => 'color: #cccccc', 'declarations' => array( 'color' => '#cccccc' ), 'classnames' => 'has-color' )`.
+ *
+ * @access public
```
Just spotted this. I don't think `@access public` is necessary, but do let me know if this is influenced by elsewhere in Core and I'm missing something

ramonjd replied:
Just to emphasize that these methods should be the public interface to the Style Engine API. 

I can't find a precedent though, so I'll remove.

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-css-declarations.php:
```diff
@@ -0,0 +1,191 @@
+ *
+ * Holds, sanitizes, processes and prints CSS declarations for the style engine.
+ *
+ * @access private
+ * @since 6.1.0
+ */
+class WP_Style_Engine_CSS_Declarations {
```
While this is intended to be private, it's open not just for instantiation, but extension.

Some thoughts:
- Should this be considered private? Why?
- If so, should this be a `final` class to at least block extension, or do we extend/plan to extend this class?
- What else could be done? (I appreciate that something like an allowlist using an optimized `debug_backtrace` for the caller of the constructor is essentially the only way to "properly" lock it down though, so not suggesting that).

Same applies to all instances in the PR.

@peterwilsoncc do you have any thoughts here?

ramonjd replied:
> or do we extend/plan to extend this class

I'm not sure of the strategy right now.

I expect that there is no plan to extend. Since this comes from a Gutenberg package, we expect to version it. 

Backwards compat will foil that I guess. 🤷 

Happy to take any advice onboard!

costdev replied:
@azaozz Do you have any thoughts on this?

azaozz replied:
@costdev thanks for the ping.

If this is intended to be private, "core use only", it will perhaps have to be locked in a closure like the current Webfonts code is. That will ensure it can be updated in the future without concerns about backwards compatibility.

If not locked, back-compat will have to be maintained, even if the class is `final`.

costdev replied:
Thanks @azaozz! _If_ it's not locked and back-compat has to be maintained, I don't think there would be any benefit in having `@access private`. Would you agree?

ramonjd replied:
> it will perhaps have to be locked in a closure like the current Webfonts code is.

Thanks for the advice here.

I think the class in question (`WP_Style_Engine_CSS_Declarations`), and the following, other, new classes:

- `WP_Style_Engine_CSS_Rule`
- `WP_Style_Engine_CSS_Rules_Store`
- `WP_Style_Engine_Processor`

should be fairly stable. Maintaining backwards compatibility will be expected and, I hope, not overly tortuous. 

I'll remove `@access private` from these.

`WP_Style_Engine`, on the other hand, is the primary integration point and provides the bulk of the functionality.

I expect it will have to be updated more frequently, especially given that this is a new API.

In that case, do folks think it's safer to refactor `WP_Style_Engine` to something similar to what's happening in webfonts? https://github.com/ramonjd/wordpress-develop/blob/bb0cbe3f557d16714ca109a28c2877d178adc94f/src/wp-includes/script-loader.php#L3059-L3058

cc @aristath and @andrewserong for thoughts

andrewserong replied:
> WP_Style_Engine, on the other hand, is the primary integration point and provides the bulk of the functionality.

I think even this one is pretty stable, and good for us to commit to maintaining for backwards compatibility. The protected methods can be added to or rearranged safely in future versions, but the public methods appear to be ones that have been pretty stable in Gutenberg lately, and I think would be good to commit to in the long-term. I'd lean toward keeping this PR as-is.

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-css-declarations.php:
```diff
@@ -0,0 +1,191 @@
+	}
+
+	/**
+	 * Add a single declaration.
```
```suggestion
	 * Adds a single declaration.
```

Active not imperative (applies to all instances in the PR) :slightly_smiling_face:

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-css-declarations.php:
```diff
@@ -0,0 +1,191 @@
+	 * @return WP_Style_Engine_CSS_Declarations Returns the object to allow chaining methods.
+	 */
+	public function add_declaration( $property, $value ) {
+
```
```suggestion
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-css-rule.php:
```diff
@@ -0,0 +1,140 @@
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return WP_Style_Engine_CSS_Declarations
```
```suggestion
	 * @return WP_Style_Engine_CSS_Declarations The declarations object.
```
Even if a method's decription indicates this, all `@return` should have a description (think for DevHub readers).

Applies to all instances in the PR

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-css-rule.php:
```diff
@@ -0,0 +1,140 @@
+		$rule_indent         = $should_prettify ? str_repeat( "\t", $indent_count ) : '';
+		$declarations_indent = $should_prettify ? $indent_count + 1 : 0;
+		$new_line            = $should_prettify ? "\n" : '';
+		$space               = $should_prettify ? ' ' : '';
```
Elsewhere in the PR, these are named `$suffix` and `$spacer` respectively. Not a big thing, but it would be good to have consistency by using whichever two names are clearer.

costdev commented on src/wp-includes/style-engine/class-wp-style-engine-processor.php:
```diff
@@ -0,0 +1,166 @@
+	/**
+	 * Adds rules to be processed.
+	 *
+	 * @since 6.1.0 a store to t
```
Typo here

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,552 @@
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string? $style_value  A single css preset value.
```
```suggestion
	 * @param string $style_value   A single CSS preset value.
```
Also, can we capitalize "CSS" in all docblocks/inline docs in the PR?

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,552 @@
+	 * @return string The slug, or empty string if not found.
+	 */
+	protected static function get_slug_from_preset_value( $style_value, $property_key ) {
+		if ( is_string( $style_value ) && str_contains( $style_value, "var:preset|{$property_key}|" ) ) {
```
This contains type checking for `$style_value` but not `$property_key`. Why's that?

ramonjd replied:
Because it's an internal value being passed, one that's contained in `BLOCK_STYLE_DEFINITIONS_METADATA` - see `'css_vars'`.  We know it's a string.

But I get that this method might be called on its on potentially, so I'll update.

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,552 @@
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string? $style_value  A single css preset value.
```
```suggestion
	 * @param string $style_value  A single css preset value.
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,552 @@
+	 * @since 6.1.0
+	 *
+	 * @param string $style_value  A single css preset value.
+	 * @param array  $css_vars The css var patterns used to generate the var string.
```
```suggestion
	 * @param string $style_value  A single css preset value.
	 * @param array  $css_vars     The css var patterns used to generate the var string.
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,552 @@
+	 * @param array $style_value      A single raw style value or css preset property from the $block_styles array.
+	 * @param array $style_definition A single style definition from BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @return array<string> An array of CSS classnames.
```
```suggestion
	 * @return string[] An array of CSS classnames.
```

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,552 @@
+					// One day, if there are no stored schemata, we could allow custom patterns or
+					// generate classnames based on other properties
+					// such as a path or a value or a prefix passed in options.
+					$classnames[] = strtr( $classname, array( '$slug' => $slug ) );
```
```suggestion
					/*
					 * Right now we expect a classname pattern to be stored in BLOCK_STYLE_DEFINITIONS_METADATA.
					 * One day, if there are no stored schemata, we could allow custom patterns or
					 * generate classnames based on other properties
					 * such as a path or a value or a prefix passed in options.
					 */
					$classnames[] = strtr( $classname, array( '$slug' => $slug ) );
```
See [Coding Standards - Inline Documentation Standards - 5.2 Multi-line Comments](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#5-2-multi-line-comments)

Applies to all instances in the PR.

costdev commented on tests/phpunit/tests/style-engine/styleEngine.php:
```diff
@@ -0,0 +1,666 @@
+ */
+class Tests_wpStyleEngine extends WP_UnitTestCase {
+	/**
+	 * Tear down after each test.
```
```suggestion
	 * Cleans up stores after each test.
```

costdev commented on tests/phpunit/tests/style-engine/styleEngine.php:
```diff
@@ -0,0 +1,666 @@
+	 * Tests generating block styles and classnames based on various manifestations of the $block_styles argument.
+	 *
+	 * @ticket 56467
+	 * @covers ::wp_style_engine_get_styles
+	 * @covers WP_Style_Engine::parse_block_styles
+	 * @covers WP_Style_Engine::compile_css
```
Quick check: While the test may lead to these three being called, is it specifically testing all three? If not, `@covers` should only be used for the specific function(s)/method(s) being tested.

Same question for all instances in the PR.

costdev commented on tests/phpunit/tests/style-engine/styleEngine.php:
```diff
@@ -0,0 +1,666 @@
+		);
+
+		$compiled_stylesheet = wp_style_engine_get_stylesheet_from_css_rules( $css_rules, array( 'prettify' => false ) );
+		$this->assertSame( '.saruman{color:white;height:100px;border-style:solid;align-self:unset;}.gandalf{color:grey;height:90px;border-style:dotted;align-self:safe center;}.radagast{color:brown;height:60px;border-style:dashed;align-self:stretch;}', $compiled_stylesheet );
```
I love this sample data :heart:

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssRulesStore.php:
```diff
@@ -0,0 +1,170 @@
+ */
+class Tests_Style_Engine_wpStyleEngineCSSRulesStore extends WP_UnitTestCase {
+	/**
+	 * Tear down after each test.
```
```suggestion
	 * Cleans up stores after each test.
```

costdev commented on tests/phpunit/tests/style-engine/wpStyleEngineCssRulesStore.php:
```diff
@@ -0,0 +1,170 @@
+		$selector       = '.haddock';
+
+		$new_fish_store->add_rule( $selector );
+		$this->assertEquals( $selector, $new_fish_store->add_rule( $selector )->get_selector(), 'Selector string of store rule does not match expected value' );
```
`assertEquals()` should only be used for objects or arrays containing objects. Otherwise, a more appropriate assertion, or `assertSame()` should be used. Applies to all instances in the PR.

- Try replacing all `assertEquals()` with `assertSame()`, then checking any failures to see if they are `object|object[]`.
- Always use the strictest possible assertion.

mukeshpanchal27 commented on src/wp-includes/style-engine/class-wp-style-engine-css-declarations.php:
```diff
@@ -0,0 +1,191 @@
+	 * @param string $value    The value to be filtered.
+	 * @param string $spacer   The spacer between the colon and the value. Defaults to an empty string.
+	 *
+	 * @return string The filtered declaration as a single string.
```
Should also mention when it return an empty string

costdev replied:
i.e. ", or an empty string."

Also, `as a single string` seems redundant as `$value` is documented as a string. Is there more to this that I'm missing?

costdev reviewed (changes requested):
Thanks for the updates @ramonjd! I've reviewed the other files as well in this one.

mukeshpanchal27 commented on tests/phpunit/tests/style-engine/styleEngine.php:
```diff
@@ -0,0 +1,666 @@
+	/**
+	 * Tests adding rules to a store and retrieving a generated stylesheet.
+	 *
+	 * @ticket 56467
+	 * @covers ::wp_style_engine_get_styles
+	 * @covers WP_Style_Engine::store_css_rule
```
Per [56070](https://core.trac.wordpress.org/ticket/56070), there should be a line break between `@ticket` and `@covers`.

mukeshpanchal27 reviewed (commented):
Left some nit pick reviews.

ramonjd committed:
Grammar, formatting, removing class_exists checks

ramonjd commented:
I'll fix up those failing tests later. Last commit was a bit rushed.

Thanks for the reviews, folks.

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,560 @@
+	 * @param array $style_value      A single raw style value or css preset property from the $block_styles array.
+	 * @param array $style_definition A single style definition from BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @return array[] An array of CSS classnames.
```
```suggestion
	 * @return string[] An array of CSS classnames.
```

ramonjd replied:
That was dumb. Thanks for catching that. Clearly I need to do more Typescript. :trollface:

costdev commented on src/wp-includes/style-engine/class-wp-style-engine.php:
```diff
@@ -0,0 +1,560 @@
+	 *     @type bool $convert_vars_to_classnames Whether to skip converting incoming CSS var patterns, e.g., `var:preset|<PRESET_TYPE>|<PRESET_SLUG>`, to var( --wp--preset--* ) values. Default `false`.
+	 * }
+	 *
+	 * @return array<string, string> An array of CSS definitions, e.g., array( "$property" => "$value" ).
```
```suggestion
	 * @return string[] An associative array of CSS definitions, e.g., array( "$property" => "$value" ).
```
Applies to all instances in the PR (watch out for alignment issues when this change is made for `@param`).

Regex Search for `string\[\](.*?)associative` in `wordpress-develop/src` to see precedent.

costdev reviewed (changes requested):
Thanks for the updates! I think these are the last round of changes from me. 🙂

As mentioned in the comments below, watch out for the `array<string, string>` to `string[] An associative array` changes causing alignment issues for `@param` annotations.

Also, the GHA failures don't seem to be test failures, but workflow failures, so it looks like your last commit was clean 🙂

ramonjd commented:
> Also, the GHA failures don't seem to be test failures, but workflow failures, so it looks like your last commit was clean

Good to know! I was on my phone so wasn't sure. I see red and 😱

ramonjd commented:
I might not be around next week. cc @andrewserong in case this PR and related PRs need any attention, though I think we should be sound 👍

costdev reviewed (approved):
LGTM pending the result of [this discussion](https://github.com/WordPress/wordpress-develop/pull/3199/files#r966604279). :+1: Thanks @ramonjd!

ramonjd committed:
Removing @access private from Style Engine utility classes

ockham commented:
> LGTM pending the result of [this discussion](https://github.com/WordPress/wordpress-develop/pull/3199/files#r966604279). 👍 Thanks @ramonjd!

That link doesn't seem to be working for me 😕 @costdev -- did you mean [this discussion](https://github.com/WordPress/wordpress-develop/pull/3199#discussion_r966604279) (about `@access private` and classes being `final`)? If so, has that been sufficiently [answered](https://github.com/WordPress/wordpress-develop/pull/3199#discussion_r967446258) by @ramonjd?

I'd love if we could land this PR soon, as it blocks a number of other Style Engine related PRs 😊
@andrewserong maybe you can help out a bit? 🙏

costdev commented:
That's right @ockham, not sure why the original link isn't working.

I think we're waiting on feedback from @aristath and @andrewserong on that topic. Not sure if @azaozz has anything more to add as well.

ramonjd commented:
> I think we're waiting on feedback from @aristath and @AndrewSong on that topic. 

For my part, I've been thinking about this, and I'm pretty confident that the function signatures and method names are fine for now. The classes have been running in Gutenberg for some time. 

Ultimately the purpose of all this code is to accept block style input and output compiled CSS. I expect there'll be updates to the body of some functions, and probably unforeseen future conditions that will require new methods, and later, deprecations, but that's the story of WordPress. 

The unknowns will remain unknowns :D 

I'd lean on @aristath's and @AndrewSong's opinion though.

andrewserong commented:
Thanks for pings!

> For my part, I've been thinking about this, and I'm pretty confident that the function signatures and method names are fine for now. The classes have been running in Gutenberg for some time.

I agree, and like you mention in the thread, maintaining backwards compatibility is expected for the style engine classes. One of the things I believe @aristath had in mind with these classes was the eventual use cases of plugins and themes, so we ultimately do want these to be stable and well-supported classes with consistent public methods. The structure and naming of the public methods went through lots of iterations in the Gutenberg plugin and appears to be pretty stable now from my perspective.

I think this PR looks in a good shape to land, personally.

aristath commented:
As mentioned above, the Style Engine classes went through a lot of iterations and discussion. What we ended up with is a hierarchy and structure that is solid, stable, makes sense, and I feel it can withstand the test of time 👍

audrasjb commented:
Committed in https://core.trac.wordpress.org/changeset/54156
